### DISCUSSION TITLE
Enquiry on concurrent record changes on same zone in powerdns

### DISCUSSION THREAD

**Discussion by mrjovi91** [poweradmin/poweradmin#402] [Q&A]
Hi Administrators,

We would like to enquire about the concurrency support on poweradmin.
When poweradmin tries to update records in a zone, understand that it will add the records followed by increment soa.

In the following scenario, will the zone be locked while its updating the zone records to maintain consistent state:-

1. Two users login to poweradmin is making zone record changes and submit near the same time.
2. One user is making changes via poweradmin and another user is making change via the pdns native api.

Hope to hear from you soon.

Thank you.

Best Regards,
Jovi

**Comment by edmondas** [comment]
I understand your need for the system to work properly when multiple users work on records at once. There are also issues with the current code base that limit the implementation of such solutions or the changes will be on a quick-and-dirty basis. I think your proposal should be integrated into the new system design.

**Comment by edmondas** [comment]
@mrjovi91 I was trying to think about how to handle it without a record lock. A check before saving seems the best option to me, otherwise you need to manage conflicts or have some kind of version control. What happens if we block a user in case of a conflict? We need to update his view to reflect the most recent changes. What examples or suggestions do you have from other software?

**Reply by mrjovi91** [reply]
Initially my thoughts was that on post of form, server side will check zone SOA value submitted in form vs zone SOA value in the backend. 
If the SOA is different in form vs backend, it means that someone else submitted a change too so the current form submit is no longer valid.
If the SOA is same in form vs backend, it means that no one else has submit a change and the current form submit is still valid.

On rejection of change request, maybe we can prompt error to user saying that "Request has expired, please try again" then redirect user back to the edit zone page showing the latest records.

This should eliminate most concurrent change problems without implementing lock on domain/record level.

Just a suggestion, let me know what you think. Cheers mate.

**Reply by edmondas** [reply]
In terms of implementation side, your model looks pretty straightforward, which can be done with current release, but I am not sure about redirecting user to same page, because he can lose all changes, but under current circumstances, maybe that could be the best option.

**Comment by edmondas** [comment]
@mrjovi91 There is one drawback to the solution you proposed: after you submit it, if there is a serial mismatch, the latest data will be loaded, and your changes will be lost.

**Reply by edmondas** [reply]
The next release will include additional setting that will allow you to test how your proposed conflict resolution works:
`$experimental_edit_conflict_resolution = 'only_latest_version';`

This warning will appear if serial is out of date in zone edit view:
![image](https://user-images.githubusercontent.com/9033112/212823646-e097bbbf-3cf9-4eb7-9dc2-e0a40aa6cd59.png)

**Comment by edmondas** [comment]
Closing this one since it shipped a while ago. Set `misc.edit_conflict_resolution` to `only_latest_version` in settings.php and a save whose SOA serial no longer matches the zone is rejected with "Request has expired, please try again". That covers both cases you described, two Poweradmin users and a change made through the PowerDNS API, since either bumps the serial.

The drawback I mentioned back then still stands: the rejected user's edits are not kept. Merging both sides is tracked in #1543. Keeping the rejected edits on the page is #1545.
